src/jobs/keys-update/keys-update.service.ts

```diff
@@ -133,51 +143,138 @@ export class KeysUpdateService {
+              stakingModuleAddress,
+              currentBlockHash,
+              prevNonce,
+              currNonce,
+            });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
+            continue;
+          }
 
+          if (this.isTooMuchDiffBetweenBlocks(prevElMeta.blockNumber, currElMeta.number)) {
+            this.logger.log('Too much difference between the blocks, start indexing', {
+              stakingModuleAddress,
+              currentBlockHash,
+            });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
             continue;
           }
 
-          await moduleInstance.update(contractModule.stakingModuleAddress, currElMeta.hash);
+          // calculate once per iteration
+          // no need to recheck each module separately
+          isReorgDetected = isReorgDetected ? true : await this.isReorgDetected(prevElMeta.blockHash, currentBlockHash);
+
+          if (isReorgDetected) {
+            this.logger.log('Reorg detected, start indexing', { stakingModuleAddress, currentBlockHash });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
+            continue;
+          }
+
+          if (
+            prevElMeta.blockNumber < currElMeta.number &&
+            (await moduleInstance.operatorsWereChanged(
+              contractModule.stakingModuleAddress,
+              prevElMeta.blockNumber + 1,
+              currElMeta.number,
+            ))
+          ) {
+            this.logger.log('Update operator events happened, need to update operators', {
+              stakingModuleAddress,
+              currentBlockHash,
+            });
+
+            await moduleInstance.updateOperators(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
+            continue;
+          }
+
+          this.logger.log('No changes have been detected in the module, indexing is not required', {
+            stakingModuleAddress,
+            currentBlockHash,
+          });
         }
+
+        // Update EL meta in db
+        await this.elMetaStorage.update({ ...currElMeta, lastChangedBlockHash });
       },
       { isolationLevel: IsolationLevel.READ_COMMITTED },
     );
 
     return currElMeta;
   }
 
+  public async isReorgDetected(prevBlockHash: string, currentBlockHash: string) {
```

Comment:
sorry, i can be wrong but i expect smth like this. compare at the end that earliest block in chain we check dont match prevBlock.hash. Maybe you could explain me why i am wrong

async isReorgDetected(prevBlockHash: string, currentBlockHash: string): Promise<boolean> {
    try {
        const prevBlock = await  this.provider.getBlock(prevBlockHash);
        const currBlock = await this.provider.getBlock(currentBlockHas);

        let block = currBlock;
        while (block.number > prevBlock.number) {
            block = await this.provider.getBlock(block.parentHash);
            if (!block) {
                throw new Error('Block not found during traversal');
            }
        }

        // Check if the block at prevBlock's position has prevBlock's hash
        return currentBlock.hash !== prevBlockHash; 
     }

Reply:
fixed